cmake-1.1 portgroup: generator support

RJVB opened the pull request:
This is a rebased version of the changes discussed in PR #26, fixing the conflict the only sure way I know. (cmake-1.1 portgroup: generator support; https://github.com/macports/macports-ports/pull/26#issuecomment-276305199)

raimue commented on _resources/port1.0/group/cmake-1.1.tcl:
```diff
@@ -34,73 +34,253 @@
+if {[tbool configure.ccache]} {
+    ui_error "Please don't use configure.ccache=yes on the commandline for port:${subport}"
+    return -code error "invalid invocation (port:${subport})"
+}
```
I do not see any need for such a check. Overriding any option from the command line is prone to fail and only meant to be used when you know what you are doing.

RJVB replied:
Here's what happens without overriding `configure.ccache`:

```
> port -v configure kf5-kruler configureccache=yes
...
--->  Configuring kf5-kruler
        (using ccache)
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:1 (project):
  The CMAKE_C_COMPILER:

    ccache /usr/bin/cc

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CC" or the CMake cache entry CMAKE_C_COMPILER to the full path to
  the compiler, or to the compiler name if it is in the PATH.


CMake Error at CMakeLists.txt:1 (project):
  The CMAKE_CXX_COMPILER:

    ccache /usr/bin/c++

  is not a full path and was not found in the PATH.

  Tell CMake where to find the compiler by setting either the environment
  variable "CXX" or the CMake cache entry CMAKE_CXX_COMPILER to the full path
  to the compiler, or to the compiler name if it is in the PATH.


--Configuring incomplete, errors occurred!
```

The same thing happens when you use `configure.ccache=yes` on the commandline, despite the override (which is apparently intended to fail in that case). I prefer to raise an error, telling the user what to do instead, since we can do that. (And yes, the message isn't ideal for that; will fix.)

As to
>only meant to be used when you know what you are doing.

I think I do ;)

raimue commented on _resources/port1.0/group/cmake-1.1.tcl:
```diff
@@ -34,73 +34,253 @@
+        return [list \
+            -DCMAKE_C_COMPILER_LAUNCHER=${prefix}/bin/ccache \
+            -DCMAKE_CXX_COMPILER_LAUNCHER=${prefix}/bin/ccache]
+    }
```
Does this actually work? Unless `configure.ccache` is set, write access to the ccache directory is prohibited by the sandbox. If ccache cannot write to `CCACHE_DIR` it will not error out, but it will also not cache anything, rendering this useless. Also, in trace mode execution of `${prefix}/bin/ccache` will be denied without declaring a dependency on the ccache port.

Support for ccache seems independent from the generators and could be split into a separate commit or be addressed in a subsequent pull request.

RJVB replied:
You mean apart from the fact that it works for me in all my big KF5 builds and that `CMAKE_<LANG>_COMPILER_LAUNCHER` is how one should configure the use of ccache according to what I found online? :) I think I did make the CCACHE_DIR writable for macports_user.
Adding a build dependency on port:ccache is trivial.

I know this PR could be split up because evidently I've not stopped evolving my personal copy waiting for PR #26 to be merged. I'm just not very motivated :-/

raimue replied:
I do not question that `CMAKE_<LANG>_COMPILER_LAUNCHER` is the correct way for cmake. Making the directory writable for the macports user is not enough as it also needs to be in the sandbox. When ccache encounters a permission problem it will not give you any error, but fail silently and not not write anything to the CCACHE_DIR. Try to empty `/opt/local/var/macports/build/.ccache/` and see if a build actually places any files there.

The dependency on ccache will be [added by base](https://github.com/macports/macports-base/blob/1a78a90fe253ce032822c9881d7c6db4247c2f69/src/port1.0/portutil.tcl#L1453-L1456) when executing the port target, so it will not show up in `port info`. I guess it was done this way as otherwise ccache would show up for almost all ports.

What we would need for this would be a way in base to add the CCACHE_DIR to the sandbox and add the ccache dependency, but not [prepend](https://github.com/macports/macports-base/blob/1a78a90fe253ce032822c9881d7c6db4247c2f69/src/port1.0/portconfigure.tcl#L791-L804) ccache to `CC` in the environment. At the moment this is both controlled by `configure.ccache yes`.

raimue commented on _resources/port1.0/group/cmake-1.1.tcl:
```diff
@@ -123,66 +303,55 @@ default configure.post_args {${worksrcpath}}
-#     if {${configure.optflags} ne ""} {
-#         configure.args-append   -DCMAKE_C_FLAGS="-DNDEBUG ${configure.cflags}" \
-#                                 -DCMAKE_CXX_FLAGS="-DNDEBUG ${configure.cxxflags}"
-#     }
```
If you delete this, the comment above ending with a colon no longer makes sense. I guess the two lines should be removed as well?

RJVB replied:
I've reworded the comment section.

raimue commented on _resources/port1.0/group/cmake-1.1.tcl:
```diff
@@ -284,14 +471,31 @@ platform darwin {
+# --enable-debug (regardless of variant) we remove it
+if {[string first "--enable-debug" ${configure.args}] > -1} {
+    configure.args-delete     --enable-debug
+}
```
The same statement already exists [above](https://github.com/macports/macports-ports/pull/1028/commits/00edba123fb5db4d278457152dd4f9acf9d7fa42#diff-1cfa208ef1d2a7a0f9ffef790a8c8088R350).

I even assume this is useless here, as it is evaluated before the rest of the Portfile. And even then, why is this necessary? Just do not add `--enable-debug` in the Portfile? I am confused.

RJVB replied:
You're right, I must have intended to move the code and forgot to remove it from the original location. As far as I can tell I inherited this morsel with the rest of cmake-1.0.tcl and I guess that in turn it may have been implemented because of the debug 1.0 PG . 

We could remove it, or rather I should probably undo the effect of that PG's current actions (setting the build type to debugFull while we already use a build type that makes cmake honour `configure.cflags` and `configure.cxxflags`).

raimue commented on _resources/port1.0/group/cmake-1.1.tcl:
```diff
@@ -34,73 +34,253 @@
+                                    cmake.install_prefix \
+                                    cmake.install_rpath \
+                                    cmake.module_path \
+                                    cmake_share_module_dir \
```
Why is this one using an underscore instead of `cmake.share_module_dir`?

Or rather, why turn this into an option instead of keeping this as a simple variable? Will there ever be a need to change this from a Portfile?

RJVB replied:
A number of variables were converted into options variables by explicit request in the original PR, or in the PR that gave rise to the current version of cmake-1.1 . I wasn't overly happy with a number of those changes but had little choice (and standardised on the approach afterwards to prevent further discussion).

This variable is the location where ports install their cmake modules. It's probably not necessary that ports ever change this, it's more intended as a read-only variable. It pre-existed in cmake-1.0.tcl and a few ports use it, hence the spelling with an underscore.

You'll notice there's also `cmake.module_path`, which can be used to specify additional locations to search for cmake modules.

RJVB commented:
Apparently you do question something, I hope it's not my sanity :P

Is this proof enough for you?

```
> /opt/local/bin/ccache --show-stats
cache directory                     /opt/local/var/macports/build/.ccache
primary config                      /Users/bertin/.ccache/ccache.conf
secondary config      (readonly)    /opt/local/etc/ccache.conf
cache hit (direct)                 15516
cache hit (preprocessed)            3989
cache miss                         17070
cache hit rate                     53.33 %
called for link                     1185
called for preprocessing             220
multiple source files                352
compile failed                        67
preprocessor error                    55
can't use precompiled header          12
bad compiler arguments              3409
autoconf compile/link                151
no input file                        170
cleanups performed                   172
files in cache                      9998
cache size                           1.8 GB
max cache size                       5.0 GB
```

That score may contain a small percentage NOT due to `port build` commands but I can guarantee that the vast majority comes from builds executed in a port's ${build.dir}. I can also guarantee that it makes a huge difference. A full KDevelop rebuild takes less than 10 minutes using ccache and anywhere between 1 and 2 hours without on the slower of my machines.

If failure is silent then there's no hurt either for those users who fail to do whatever additional set-up may be required (which I agree we should document). In addition, the check and option override will prevent failures, which could be a boon if you launch multiple builds in a single command.

I *am* open to removing the specific cmake.ccache option though. It doesn't seem to be required anymore, at least not visibly as something for users to control. It should also be possible to test CCACHE_DIR writability from within `cmake::ccaching`, and raise a warning if that test fails.

pmetzger commented:
Okay, as it stands, this conflicts with what's in the master branch.

What do we need to do to get this progressed or closed out?

RJVB commented:
Well, I think I messed up brilliantly trying to get the conflict resolved...

pmetzger commented:
Okay, so, @RJVB and @raimue, are we ready to merge this?

pmetzger commented:
@raimue Thanks for your attention to this! It's appreciated!

RJVB commented on _resources/port1.0/group/cmake-1.1.tcl:
```diff
@@ -34,80 +34,246 @@
+                }
+            }
+            default {
+                ui_msg "generator fallback"
```
In practice it doesn't, which is the only reason why the statement survived :)

RJVB commented:
I'm following Rainer's suggestion for a blacklist which is indeed a much nicer way than the kludgy protection I'd thrown together (probably in haste).

It wasn't clear if the suggestion was also to get rid of the entire `cmake::handle_generator` hook function. I kept it therefore; now that I've gotten used to using them I find it a bit nicer than moving the payload into the pre-configure. (I'd still need to do the same checks and actions, and the pre-configure is cluttered enough as it is already.)

pmetzger commented:
@RJVB @raimue Do we feel this is now ready to merge?

@RJVB You should probably squash the commits.

RJVB commented:
> @RJVB You should probably squash the commits.

Yeah, but which/how?

There have been so many commits to the current cmake-1.1 version since I created the initial version of this PR ...

pmetzger commented:
There are no conflicts with a merge right now, so just the two you have now.

RJVB commented:
> There are no conflicts with a merge right now, so just the two you have now.

Is it normal that I had to force-push to get around

```
Pushing to ssh://git@github.com/RJVB/macfork-ports
To ssh://github.com/RJVB/macfork-ports
 = [up to date]            master -> master
 ! [rejected]              cmake11PG -> cmake11PG (non-fast-forward)
updating local tracking ref 'refs/remotes/origin/master'
error: failed to push some refs to 'ssh://git@github.com/RJVB/macfork-ports'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

(after doing a git rebase -i <hash> where I did a fixup of the 2nd commit)

Seems it didn't lead to problems...

pmetzger commented:
Are the changes currently reflected in the file as seen in here in the web interface to github the ones you are expecting to go in? If so, we can merge.

RJVB commented:
>Are the changes currently reflected in the file as seen in here in the web interface to github the ones you are expecting to go in? If so, we can merge. 



I think so, and diff agrees when I click on "commits / view file / raw", download the file and compare it to what I have on disk.

pmetzger commented:
Okay, @raimue, let me know if you have an objections to my committing these changes.

raimue commented:
No, no, I did not mean to get rid of `cmake::handle_generator` completely. Just no longer apply the changes every time the option is read. The current form is just fine.

I was just a bit curious why you have that check not to error out when running from the registry. Both Portfile and group file are stored in the registry, so there is no chance that an unknown value would ever make it there. But this can still be addressed later.

The changes can be merged (after squashing).

RJVB commented:
> I was just a bit curious why you have that check not to error out when running from the registry.

I honestly cannot remember. Two possibilities:
- I encountered a situation where raising a hard error when running from the registry led to all kinds of woes, and added the check
- Idem, but with a different PortGroup, and I added the check to the CMake PG too, to be safe.

I do remember having to dig out the PortGroup copy in the registry and add this kind of check to it, to unblock the situation I'd gotten myself into.

BTW, does each installed port have its own copy of *all* PortGroup files it includes, even those included by other PGs? That's the case with the cmake PG.

raimue commented:
> BTW, does each installed port have its own copy of *all* PortGroup files it includes, even those included by other PGs? That's the case with the cmake PG.

Yes, the list of used port groups is gathered in the implementation of the `PortGroup` statement, so it will work recursively.

pmetzger commented:
Okay, I'll merge this stuff tomorrow morning.

RJVB commented:
>Yes, the list of used port groups is gathered in the implementation of the `PortGroup` statement, so it will work recursively.

```
lappend PortInfo(portgroups) [list $group $version]
```

Looks like it is indeed. Strange, I'm pretty certain I've run into situations where this didn't work as expected.

pmetzger commented:
@RJVB Please let me know when you feel this is ready to commit...

RJVB commented:
> @RJVB Please let me know when you feel this is ready to commit...

Heh, it's not my habit to put up PRs for things I feel aren't ready, so yes (and sorry for not confirming this sooner) ;)

pmetzger commented:
Merged.
